File: packages/sdk-android/rulemind-core/src/main/kotlin/com/rulemind/core/ExperimentManager.kt
package com.rulemind.core

import com.rulemind.core.models.Experiment
import com.rulemind.core.models.ExperimentVariant
import java.math.BigInteger
import java.security.MessageDigest

class ExperimentManager {
    fun assign(experiment: Experiment, userId: String?): ExperimentVariant? {
        if (experiment.status != "running" || userId.isNullOrBlank() || experiment.variants.isEmpty()) {
            return null
        }
        val digest = MessageDigest.getInstance("MD5").digest("${experiment.id}:$userId".toByteArray())
        val bucket = BigInteger(1, digest).mod(BigInteger.valueOf(100)).toInt()
        var cumulative = 0
        for (variant in experiment.variants) {
            cumulative += variant.weight
            if (bucket < cumulative) {
                return variant
            }
        }
        return experiment.variants.last()
    }
}

File: packages/sdk-flutter/lib/src/experiment_manager.dart
import "dart:convert";

import "package:crypto/crypto.dart";

import "models.dart";

class ExperimentManager {
  ExperimentVariant? assign(Experiment experiment, String? userId) {
    if (experiment.status != "running" || userId == null || userId.isEmpty || experiment.variants.isEmpty) {
      return null;
    }
    final digest = md5.convert(utf8.encode("${experiment.id}:$userId")).toString();
    final bucket = BigInt.parse(digest, radix: 16) % BigInt.from(100);
    var cumulative = 0;
    for (final variant in experiment.variants) {
      cumulative += variant.weight;
      if (bucket.toInt() < cumulative) {
        return variant;
      }
    }
    return experiment.variants.last;
  }
}
